### DISCUSSION TITLE
A2DP Source and I²S slave

### DISCUSSION THREAD

**Discussion by g3gg0** [pschatzmann/ESP32-A2DP#329] [Q&A] [answered]
Hi, 

great project. Really appreciate your work put into making a proper library.

I am currently planning to get a toniebox bluetooth ready for xmas (internally it has CC3200 - > DAC3100 via I²S).
I could not find something like that implemented in your lib already, and I will probably have to implement it on my own. 

Hardware will most likely be a XIAO ESP32C3, eavesdropping DIN, BCLK and WCLK. 

Q: Did you already experiment with such operating mode?
Q: As the clock master will be an external chip (unavoidable), do you think i would have to resample dynamically to not get A2DP stream out of sync (buffer under/overflow)? Or is this already covered by the internal encoder?

Thank you. 

Regards,
g3gg0

**Comment by pschatzmann** [comment, marked as answer]
It is not quite clear to me what exactly you try to do. 
I just want to point out that the ESP32C3 does not support A2DP as you can see in the Wiki!
I never used a ESP32 I2S in slave mode, but you just need to use I2S_MODE_SLAVE instead of I2S_MODE_MASTER in i2s_config.
It is already quite difficult to implement a reliable A2DP Source. If you plan to feed it via I2S I would expect that this might get quite challenging, since you need to make sure that the data is available when the callback is executed!

**Reply by g3gg0** [reply]
Thanks for your reply.

Interesting that they are varying features like hell now. Was more or less using [this ](https://docs.espressif.com/projects/esp-idf/en/v4.3/esp32c3/hw-reference/chip-series-comparison.html) and some [other ](https://docs.espressif.com/projects/esp-idf/en/v4.3/esp32c3/api-reference/bluetooth/esp_a2dp.html)sources, leading me into thinking, it has the normal bluetooth stack. Well, then i'd choose one of the older ESPs, i have enough of them.
Didn't see the Wiki, thanks for pointing out.

Use case is an existing toy audio player (Toniebox) which uses CC3200, I²S and DAC3100 to decode and play .ogg audio files.
This box has no bluetooth support and I wanted it to implement the elegant way by eavesdropping I²S.
DAC3200 seems to be driven as slave and the ESP32 shall sit there in parallel.

Yeah from your reply i read, that the A2DP clock (ruling when how much data must be available via that callback) cannot vary, 
but is determined by the peripheral clock and software stack behind with no variance allowed.
So i must prebuffer I²S captured data and dynamically resample it in that callback to match clock deviation and also taking care of the clock variances.

Here a small visualization:
![grafik](https://user-images.githubusercontent.com/479231/204792884-af359fb7-2129-4c92-95c5-2c87eed9bef0.png)


Regards,
g3gg0

**Reply by pschatzmann** [reply]
In my [audiotools](https://github.com/pschatzmann/arduino-audio-tools) I am providing some [fixed resampling](https://github.com/pschatzmann/arduino-audio-tools/wiki/Resampling). 

It would be quite interesting if we could provide some variable resampling as well.

**Reply by g3gg0** [reply]
I have implemented it [in C#](https://github.com/g3gg0/rx-fft/blob/master/EZ-USB/RX-FFT/LibRXFFT/Libraries/SignalProcessing/Resampler.cs) years ago, but not sure how good it was.
I think it was not perfectly accurate, as I didn't save rounding errors. But yeah, will probably have to look into it again.

Regards,
g3gg0
